File: java/test/TestErrors.java
import mpc.MPCDesignation;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

/**
 * Test MPC designation error handling.
 */
public class TestErrors {

    /**
     * Convert escape sequences in test input.
     */
    private static String unescapeInput(String input) {
        String result = input;
        result = result.replace("\\t", "\t");
        result = result.replace("\\n", "\n");
        result = result.replace("\\r", "\r");
        result = result.replace("\\f", "\f");
        result = result.replace("\\v", "\u000B");

        // Handle \x00 style escapes
        StringBuilder sb = new StringBuilder();
        int i = 0;
        while (i < result.length()) {
            if (i + 3 < result.length() && result.charAt(i) == '\\' && result.charAt(i + 1) == 'x') {
                try {
                    String hexStr = result.substring(i + 2, i + 4);
                    int val = Integer.parseInt(hexStr, 16);
                    sb.append((char) val);
                    i += 4;
                } catch (NumberFormatException e) {
                    sb.append(result.charAt(i));
                    i++;
                }
            } else {
                sb.append(result.charAt(i));
                i++;
            }
        }

        return sb.toString();
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: TestErrors <error_test_cases.csv>");
            System.exit(1);
        }

        String csvFile = args[0];
        int totalTests = 0;
        int passedTests = 0;
        int failedTests = 0;

        System.out.println("=== MPC Designation Error Tests (Java) ===");
        System.out.println();

        try (BufferedReader reader = new BufferedReader(new FileReader(csvFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                // Skip empty lines, comments, and header
                if (line.isEmpty() || line.startsWith("#") || line.startsWith("category,")) {
                    continue;
                }

                // Parse CSV line
                String[] parts = line.split(",", 5);
                if (parts.length < 5) {
                    continue;
                }

                String category = parts[0];
                String subcategory = parts[1];
                String rawInput = parts[2];
                String expectedError = parts[3];
                String description = parts[4];

                String input = unescapeInput(rawInput);
                totalTests++;

                boolean gotError = false;
                String errorType = "";
                String resultOutput = "";

                try {
                    MPCDesignation.Result result = MPCDesignation.convert(input);
                    resultOutput = result.output;
                } catch (MPCDesignation.MPCDesignationException e) {
                    gotError = true;
                    String errStr = e.getMessage();
                    if (errStr.contains("out of range") || errStr.contains("Invalid asteroid number")) {
                        errorType = "range";
                    } else {
                        errorType = "format";
                    }
                }

                String testId = category + "/" + subcategory;

                if (expectedError.equals("valid")) {
                    // Should succeed
                    if (gotError) {
                        failedTests++;
                        System.out.println("FAIL [" + testId + "]: '" + description + "'");
                        System.out.println("      Expected: valid conversion");
                        System.out.println("      Got:      error (" + errorType + ")");
                    } else {
                        passedTests++;
                    }
                } else {
                    // Should fail
                    if (!gotError) {
                        failedTests++;
                        System.out.println("FAIL [" + testId + "]: '" + description + "'");
                        System.out.println("      Expected: error (" + expectedError + ")");
                        System.out.println("      Got:      '" + resultOutput + "' (success)");
                    } else {
                        passedTests++;
                    }
                }
            }
        } catch (IOException e) {
            System.err.println("Cannot open file: " + csvFile);
            System.exit(1);
        }

        System.out.println();
        System.out.println("=== Error Test Results ===");
        System.out.println("Total:  " + totalTests);
        System.out.println("Passed: " + passedTests);
        System.out.println("Failed: " + failedTests);

        if (failedTests > 0) {
            System.exit(1);
        }
    }
}

File: kotlin/test/TestErrors.kt
import mpc.MPCDesignation
import java.io.File

/**
 * Test MPC designation error handling.
 */
fun main(args: Array<String>) {
    if (args.isEmpty()) {
        System.err.println("Usage: TestErrors <error_test_cases.csv>")
        System.exit(1)
    }

    val csvFile = args[0]
    var totalTests = 0
    var passedTests = 0
    var failedTests = 0

    println("=== MPC Designation Error Tests (Kotlin) ===")
    println()

    try {
        File(csvFile).useLines { lines ->
            for (line in lines) {
                // Skip empty lines, comments, and header
                if (line.isEmpty() || line.startsWith("#") || line.startsWith("category,")) {
                    continue
                }

                // Parse CSV line
                val parts = line.split(",", limit = 5)
                if (parts.size < 5) continue

                val category = parts[0]
                val subcategory = parts[1]
                val rawInput = parts[2]
                val expectedError = parts[3]
                val description = parts[4]

                val input = unescapeInput(rawInput)
                totalTests++

                var gotError = false
                var errorType = ""
                var resultOutput = ""

                try {
                    val result = MPCDesignation.convert(input)
                    resultOutput = result.output
                } catch (e: MPCDesignation.MPCDesignationException) {
                    gotError = true
                    val errStr = e.message ?: ""
                    errorType = if (errStr.contains("out of range") || errStr.contains("Invalid asteroid number")) {
                        "range"
                    } else {
                        "format"
                    }
                }

                val testId = "$category/$subcategory"

                if (expectedError == "valid") {
                    // Should succeed
                    if (gotError) {
                        failedTests++
                        println("FAIL [$testId]: '$description'")
                        println("      Expected: valid conversion")
                        println("      Got:      error ($errorType)")
                    } else {
                        passedTests++
                    }
                } else {
                    // Should fail
                    if (!gotError) {
                        failedTests++
                        println("FAIL [$testId]: '$description'")
                        println("      Expected: error ($expectedError)")
                        println("      Got:      '$resultOutput' (success)")
                    } else {
                        passedTests++
                    }
                }
            }
        }
    } catch (e: Exception) {
        System.err.println("Cannot open file: $csvFile")
        System.exit(1)
    }

    println()
    println("=== Error Test Results ===")
    println("Total:  $totalTests")
    println("Passed: $passedTests")
    println("Failed: $failedTests")

    if (failedTests > 0) {
        System.exit(1)
    }
}

/**
 * Convert escape sequences in test input.
 */
private fun unescapeInput(input: String): String {
    var result = input
        .replace("\\t", "\t")
        .replace("\\n", "\n")
        .replace("\\r", "\r")
        .replace("\\f", "\u000C")
        .replace("\\v", "\u000B")

    // Handle \x00 style escapes
    val sb = StringBuilder()
    var i = 0
    while (i < result.length) {
        if (i + 3 < result.length && result[i] == '\\' && result[i + 1] == 'x') {
            try {
                val hexStr = result.substring(i + 2, i + 4)
                val value = hexStr.toInt(16)
                sb.append(value.toChar())
                i += 4
            } catch (e: NumberFormatException) {
                sb.append(result[i])
                i++
            }
        } else {
            sb.append(result[i])
            i++
        }
    }

    return sb.toString()
}
